<a href="https://colab.research.google.com/github/Adeela-saleem/MachineLearning_Algorithms/blob/main/Regression/Multiple_LR.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from sklearn.datasets import make_regression

In [2]:
import pandas as pd
import numpy as np


In [4]:
X,y=make_regression(n_samples=100,n_features=2,n_informative=2,n_targets=1,noise=50)

In [6]:
df=pd.DataFrame({'feature1':X[:,0],'feature2':X[:,1],'target':y})

In [7]:
df.shape

(100, 3)

In [8]:
df.head()

,feature1,feature2,target
0,0.427816,0.172064,38.381585
1,-1.819968,0.299065,-234.627746
2,1.166364,-0.070811,174.985062
3,-0.786997,0.010803,37.370005
4,0.607807,-0.010466,118.806759


In [20]:
import plotly.express as px



In [19]:
fig=px.scatter(df,x='feature1',y='feature2',color='target')
fig.show()

In [11]:
from sklearn.model_selection import train_test_split

In [12]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [13]:
from sklearn.linear_model import LinearRegression

In [14]:
lr=LinearRegression()

In [15]:
lr.fit(X_train,y_train)

LinearRegression()

In [16]:
y_pred=lr.predict(X_test)

In [17]:
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score


In [21]:

print("MAE",mean_absolute_error(y_test,y_pred))
print("MSE",mean_squared_error(y_test,y_pred))
print("R2 score",r2_score(y_test,y_pred))

MAE 36.296000344814786
MSE 2558.632270804919
R2 score 0.812916740769758


Here is a concise summary of our entire conversation:

---

### **1. Plotly Overview & `px` vs. `go**`

* **`plotly.express` (`px`):** A high-level, easy-to-use library for quickly generating interactive charts (zoom, hover, filter) with minimal code.
* **`plotly.graph_objects` (`go`):** A low-level API offering granular control over layouts, trace overlays, 3D surface plots, and custom dashboard components.
* **When to Use Plotly:** Ideal for multi-dimensional data, interactive web/notebook dashboards, financial charts, and map visualizations.

---

### **2. Scatter Plot Observation**

* **Visualization:** Plotted `feature1` ($X$-axis) against `feature2` ($Y$-axis) with a color gradient representing the continuous `target` variable.
* **Insight:** Displays a smooth transition from lower values (bottom-left) to higher values (top-right), confirming a clear linear relationship suitable for **Multiple Linear Regression**.

---

### **3. Model Performance & Evaluation Metrics**

* **$R^2$ Score (`0.813`):** Explains **~81.3%** of the variance in the target variable, indicating a strong fit.
* **MAE (`36.30`):** Predictions off by an average of 36.30 units.
* **MSE (`2558.63`) & RMSE (`~50.58`):** Measures penalizing larger errors caused by synthetic noise (`noise=50`).